src/main.c

```diff
@@ -486,10 +487,20 @@ static void async_nb_on_write_complete(h2o_socket_t *sock, const char *err)
     async_nb_submit_write_pending();
 }
 
+#define NEVERBLEED_MAX_IN_FLIGHT_TX 200
```

Comment:
maybe this should be a configurable option, off by default?

Or set dynamically via retrieving the unix socket buffer size / max size of transaction?

Reply:
Am I correct to understand that when neverbleed writes a partial response and h2o reads it, h2o will block until neverbleeds writes the rest of the response?

Assuming that is the case (that it will not be a busy loop), I think we would be fine with a hard-coded value.

In reality, unless hardware offload is used, h2o will be faster in reading than neverbleed writing, so this setting matters probably only to us.

Regarding the choice of 200, by looking at the "socket(2) OS comparison" table of https://www.netmeister.org/blog/ipcbufs.html (see "PF_LOCAL stream"), I think the value will be fine on Linux, as the default buffer size is ~200KB and the maximum number of chunks the buffer can store is >=278. Though, considering the operating systems with buffers as small as 8KB, probably a slightly smaller value might be considered more portable (but as stated above, I do not think it matters in practice).

Reply:
> Am I correct to understand that when neverbleed writes a partial response and h2o reads it, h2o will block until neverbleeds writes the rest of the response?

This is the case when using `async_nb_run_sync`, however, I am proposing a sort of busy loop on the async read side: "read transactions until EAGAIN". We can do this with the assumption that neverbleed will always atomically write a response (a partial write should never happen as we're limiting the # of tx in-flight and so the buffer will never be full), therefore when a socket is read ready, there will always be a full response to read

Reply:
Thanks.

I think I agree that having a busy loop might not be that bad, though I'm not sure if there is a reason to switch from using a blocking read (that helps neverbleed make progress compared to a busy loop)?

Rather than adding a new code path that runs the transactions, I wonder if we could just repeat calling `async_nb_run_sync` inside `async_nb_read_ready` until `ioctl(FIONREAD)` says that the UNIX socket receive bufer is empty.